### DISCUSSION TITLE
model.export() not terminating

### DISCUSSION THREAD

**Discussion by tr1stn** [MPh-py/MPh#195] [General]
I'd like to export several images from a model. In order to do this, I defined the model, multiple plot nodes and export nodes for each plot. Solving the model and exporting the images through the GUI works without problems.

Now I'd like to automate this process and wrote this script:
```python
import argparse
import mph

parser = argparse.ArgumentParser()
parser.add_argument('filename')
args = parser.parse_args()

client = mph.start()
model = client.load(args.filename)
model.solve()
print("Model opened and solved")

print("Exporting all images")
model.export()
print("Images exported")

exit()
```

My problem: model.export() doesn't terminate. Interrupting it with Ctrl+C gives the following output:
```
Exporting all images
^CTraceback (most recent call last):
  File "SourceFile", line 72, in com.comsol.clientapi.impl.ExportFeatureClient.run
  File "SourceFile", line 270, in com.comsol.clientapi.engine.APIEngine.runMethod
  File "SourceFile", line 290, in com.comsol.clientapi.engine.APIEngine.runMethod
Exception: Java Exception

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "export_images_from_comsol.py", line 42, in <module>
    model.export()
  File "/path/to/python/venv/lib64/python3.6/site-packages/mph/model.py", line 803, in export
    node.run()
  File "/path/to/python/venv/lib64/python3.6/site-packages/mph/node.py", line 540, in run
    java.run()
java.lang.NullPointerException: java.lang.NullPointerException
Error while disconnecting client at session clean-up.
Traceback (most recent call last):
  File "/path/to/python/venv/lib64/python3.6/site-packages/mph/session.py", line 154, in cleanup
    client.disconnect()
  File "/path/to/python/venv/lib64/python3.6/site-packages/mph/client.py", line 439, in disconnect
    self.java.disconnect()
RuntimeError: java.lang.IllegalStateException: Tried to disconnect while not connected.

``` 

I am using a comsol installation on a remote system. To me, this error looks as if there are some issues with the java runtime. If so, can I find debug information of model.export() somewhere? I'd also appreciate any other thoughts on that problem :)

Thanks a lot!

**Comment by john-hen** [comment]
Hi. Yes, I've also noticed that. And the tl;dr is, unfortunately, that I don't know why that happens, how to fix it, nor really how to debug it.

It's briefly mentioned in the documentation, in "Installation":
> Linux support has only been rigorously tested on Ubuntu. Occasional problems may occur on other distributions. Specifically with image exports, for which Comsol depends on external libraries.

And that's actually a bit outdated. When I originally tested this on an Ubuntu workstation, it worked. That is, the entire test suite passed, including image exports. Then I tested on an HPC cluster I had access to, with CentOS 7 as the operating system. There the image export would hang, just like you describe. So I figured it's the Linux distribution's fault. But not long ago, that same cluster was upgraded to Ubuntu 22, and there the image export also hangs, despite it being Ubuntu.

I don't know what's up with that. I've certainly tried putting various libraries on `LD_LIBRARY_PATH`, to no avail. But haven't dug much deeper. It's obviously nothing in the Python code. After all, the same Comsol API call works fine on Windows, macOS, and *some* Linux installations. I've seen similarly inexplicable errors with image exports, though in that case outright crashes, when I tried to run the Comsol client in "stand-alone" mode on Linux. Which requires putting a bunch of stuff on `LD_LIBRARY_PATH`, hence this being my first impulse for "client-server" mode too.

I guess the next thing to try would be to cut out the Python/JPype middleman. That is, translate that Python script to Java, compile and run it, while having a separate Comsol server process already up and running for the client to connect to. I could assist with that if you wanna try. There's also an issue/discussion somewhere with Java code and build instructions, though I can't remember which one off the top of my head. But the question will be, what then? If the same behavior occurs, you could report it to Comsol. If it then works, then no clue what to do about it. It might point to an issue with JPype, but since Comsol isn't open source, that's probably the end of that. Unless there is something we've forgotten to set up in the enviroment, like some search path.

**Comment by tr1stn** [comment]
Thanks a lot! The cluster I'm working on is running CentOS, so that fits well with your description. I'll try to get a windows PC with a local Comsol installation and run my script there. That's probably the easiest way…

**Comment by john-hen** [comment]
As far as I can tell, this issue has been fixed in Comsol 6.3. On the Ubuntu cluster I have access to, the image export hangs with Comsol 6.2, but not with 6.3.
